simapp/sim_test.go

```diff
@@ -95,30 +103,137 @@ func appStateFromGenesisFileFn(r *rand.Rand, accs []simulation.Account, genesisT
-func appStateRandomizedFn(r *rand.Rand, accs []simulation.Account, genesisTimestamp time.Time,
+func appStateRandomizedFn(
+	r *rand.Rand, accs []simulation.Account, genesisTimestamp time.Time,
 ) (json.RawMessage, []simulation.Account, string) {
 
-	var genesisAccounts []genaccounts.GenesisAccount
-	genesisState := NewDefaultGenesisState()
 	cdc := MakeCodec()
+	appParams := make(simulation.AppParams)
+	genesisState := NewDefaultGenesisState()
+
+	if paramsFile != "" {
+		bz, err := ioutil.ReadFile(paramsFile)
+		if err != nil {
+			panic(err)
+		}
+
+		cdc.MustUnmarshalJSON(bz, &appParams)
+	}
+
+	var (
+		genesisAccounts    []genaccounts.GenesisAccount
+		amount             int64
+		numInitiallyBonded int64
+	)
+
+	appParams.GetOrGenerate(cdc, "stake_per_account", &amount, r, func(r *rand.Rand) { amount = int64(r.Intn(1e12)) })
+	appParams.GetOrGenerate(cdc, "initially_bonded_validators", &amount, r, func(r *rand.Rand) { numInitiallyBonded = int64(r.Intn(250)) })
 
-	amount := int64(r.Intn(1e12))
-	numInitiallyBonded := int64(r.Intn(250))
 	numAccs := int64(len(accs))
 	if numInitiallyBonded > numAccs {
 		numInitiallyBonded = numAccs
 	}
-	fmt.Printf("Selected randomly generated parameters for simulated genesis:\n"+
-		"\t{amount of stake per account: %v, initially bonded validators: %v}\n",
-		amount, numInitiallyBonded)
 
+	fmt.Printf(
+		`Selected randomly generated parameters for simulated genesis:
+{
+  stake_per_account: "%v",
+  initially_bonded_validators: "%v"
+}
+`, amount, numInitiallyBonded,
+	)
+
+	genGenesisAccounts(cdc, r, accs, genesisTimestamp, amount, numInitiallyBonded, genesisAccounts, genesisState)
+	genAuthGenesisState(cdc, r, appParams, genesisState)
+	genBankGenesisState(cdc, r, appParams, genesisState)
+	genGovGenesisState(cdc, r, appParams, genesisState)
+	genMintGenesisState(cdc, r, appParams, genesisState)
+	genDistrGenesisState(cdc, r, appParams, genesisState)
+	stakingGen := genStakingGenesisState(cdc, r, accs, amount, numAccs, numInitiallyBonded, appParams, genesisState)
+	genSlashingGenesisState(cdc, r, stakingGen, appParams, genesisState)
+
+	appState, err := MakeCodec().MarshalJSON(genesisState)
+	if err != nil {
+		panic(err)
+	}
+
+	return appState, accs, "simulation"
+}
+
+func appStateFn(
+	r *rand.Rand, accs []simulation.Account, genesisTimestamp time.Time,
+) (json.RawMessage, []simulation.Account, string) {
+
+	if genesisFile != "" {
+		return appStateFromGenesisFileFn(r, accs, genesisTimestamp)
+	}
+
+	return appStateRandomizedFn(r, accs, genesisTimestamp)
+}
+
+func genAuthGenesisState(cdc *codec.Codec, r *rand.Rand, ap simulation.AppParams, genesisState map[string]json.RawMessage) {
+	authGenesis := auth.NewGenesisState(
+		nil,
+		auth.NewParams(
+			func(r *rand.Rand) uint64 {
+				var v uint64
+				ap.GetOrGenerate(cdc, "max_memo_characters", &v, r, func(r *rand.Rand) { v = simulation.ModuleParamSimulator["max_memo_characters"](r).(uint64) })
+				return v
+			}(r),
```

Comment:
This code structure is quite obfuscated, I understand the thinking to compress all the code required to getOrGen a param into an immediately executed function as is done here. But I think a more intuitive/digestable structure would be to just individually get all the parameters at the top of this function (before defining `authGenesis`). Something like: 

``` golang
func genAuthGenesisState(cdc *codec.Codec, r *rand.Rand, ap simulation.AppParams, genesisState map[string]json.RawMessage) {

     var maxMemoChars uint64
     ap.GetOrGenerate(cdc, "max_memo_characters", &maxMemoChars, r, func(r *rand.Rand) { v = simulation.ModuleParamSimulator["max_memo_characters"](r).(uint64) })
     ....

     authGenesis := auth.NewGenesisState(
		nil,
		auth.NewParams(
			maxMemoChars,
                         ...
```

This also applies too all modules aka `genXXXXGenesisState`

Reply:
Yeah so I understand your concern upon first reading the code. I think the additional verbosity that results from such a suggested change isn't worth it because now that we have constants that are directly mapped to the meaning of the arguments, it's very clear what it's doing and what value it provides.

I'd rather not have to go through each of these for what will probably be a tiny (if any) gain in brevity.

Reply:
So the code would actually become less verbose under the suggested change, however we would have to come up with variable names - that's pretty much the only downside. I don't mind  quickly making these changes if it's just a matter of the work